### ISSUE TITLE
add calendar exports for upcoming maintenance windows

### ISSUE THREAD

**Issue by anarcat** [cstate/cstate#329]
**Is your feature request related to a problem? Please describe.**
We have an issue where we have to duplicate information between cstate and ... well, a bunch of things. :) It's nice that cstate is separate, so we expect some of that, but one in particular (maintenance window times) feels like a low-hanging fruit that might be relatively easy to fix.

**Describe the solution you'd like**
On top of generating HTML and RSS feeds, I think it would be pretty neat to have events rendered as calendar files. While the standardization of that file format is [surprisingly complicated historically](https://en.wikipedia.org/wiki/ICalendar), for simple events like this it should be relatively simple to create a calendar file that people could just add to their calendar.

The idea would be to show events, especially upcoming ones, so that people can be warned of coming maintenances.

**Describe alternatives you've considered**
I *think* it might be possible to shove the RSS feed in some sort of filter to turn it into an ical file we could use.

**Additional context**

Thanks!

**Comment by mistermantas** [comment]
Hi! Thanks for your suggestion, I appreciate your continued use of the project.

Calendar files from what I understand are for one time events and adding those should be relatively easy.

Are you looking to add multiple events in one click? I think that might be a bit more difficult to accomplish, although I have not tried it or looked at the technical specifications of how that would work—but I’ve not seen that as even an option on most consumer websites.

also, it’s worth considering if there is an easier way to accomplish this use case with an external tool as at the moment this would not be a high priority feature to add - but certainly could be a cool addition once I have some more free time

**Comment by anarcat** [comment]
> Hi! Thanks for your suggestion, I appreciate your continued use of the project.

You're welcome! Thanks for maintaining cstate! :) It's been humming
along fine for years for us at this point, it's great!

> Calendar files from what I understand are for one time events and adding those should be relatively easy.

Well, calendar files are complicated beasts. They can have recurring
events, for example...

> Are you looking to add multiple events in one click?

"click"? I post events on the site by adding markdown files in a git
repo, no clicks involved. :)

> I think that might be a bit more difficult to accomplish, although I have not tried it or looked at the technical specifications of how that would work—but I’ve not seen that as even an option on most consumer websites.

Well, this is why i compare this to the RSS feed, as it's a similar
pattern...

Normally, when I post an event to cstate, i create a file in
content/issues/2024-11-04-foo.md, which gets turned into a
issues/2024-11-04-foo.html file on the website. But it also gets
aggregated on the frontpage, issues listing and the RSS feed. That RSS
feed is an entirely different format than the other stuff, and is
basically a sorted list of metadata about the posts.

I figured the implementation of a calendar file would be similar to the
RSS feed, but I haven't looked very closely at how cstate is
implemented. I suspect that hugo is the one doing the grunt work for RSS
and that to get calendars implemented in cstate, one would need to look
first at plugins in hugo...

And perhaps, in that sense, it's something that would better live
outside of cstate... But it would sure be nice to at least have some
docs on how to do this!

In fact, searching around the web, it looks like there's already a
"calendar" output format in hugo, at least according to:

https://blog.claude.nl/posts/hugo-recurring-events-with-icalendar-support/

This page also mentions "calendar" as a supported output format:

https://gohugo.io/templates/output-formats/

So it might be as simple as adding "calendar" to the "outputs" list in
the configuration, and linking to it in the base template
somehow... I'll see if i can experiment with this myself.

> also, it’s worth considering if there is an easier way to accomplish this use case with an external tool as at the moment this would not be a high priority feature to add - but certainly could be a cool addition once I have some more free time

Of course! As I said, it's probably possible to write an external thing
that would parse RSS and turn that into ICS but my hunch is that this
would be relatively simple to do in Hugo directly... Here are other
example implementations:

https://discourse.gohugo.io/t/create-one-ics-file-that-contains-all-events/42014/3
https://discourse.gohugo.io/t/ical-recurring-events/39624/6
https://discourse.gohugo.io/t/calendar-ics-template/7137/9

**Comment by mistermantas** [comment]
> "click"? I post events on the site by adding markdown files in a git
> repo, no clicks involved. :)

To clarify: I was talking about front-end for users/consumers which can expose a "Add this event to your calendar" button or similar.

Did I misunderstand and you're looking for a command line tool? The outputs solution would be great if it can work as expected as well.

**Comment by anarcat** [comment]
> To clarify: I was talking about front-end for users/consumers which can expose a &quot;Add this event to your calendar&quot; button or similar.

Right, in my mind that would just be a link to the ICS file, with a
simple `<a href>`.

> Did I misunderstand and you&#39;re looking for a command line tool?

No no, just something on the website is fine.

> The outputs solution would be great if it can work as expected as well.

Yeah, I really need to find the time to just test it!

**Comment by mistermantas** [comment]
If you find a solution that works for you, @anarcat - please feel free to share it here in this ticket. I'll also investigate later once I'm more free to work on adding new features to cstate  :)

**Comment by anarcat** [comment]
so i tried just adding "calendar" to the "outputs" and that's not
enough, i need a template too:

```
WARN 2024/11/04 09:39:04 found no layout file for "Calendar" for kind "home": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.
```

i need to move onto something else, but i might try again later.

**Comment by mistermantas** [comment]
I've added a sort of beta/alpha implementation of this - because descriptions don't seem to format properly and the end time doesn't seem to line up properly

https://github.com/cstate/cstate/blob/ad0c67039be177e60dade470a4c5d8ffe117ba98/layouts/issues/issue.html

**Comment by anarcat** [comment]
i think i don't understand this fix: looking at [ad0c670](https://github.com/cstate/cstate/commit/ad0c67039be177e60dade470a4c5d8ffe117ba98), it refers to this issue, but doesn't seem related to the calendar stuff...

i've also tried adding a "calendar" output to the "home" output and it didn't change much... see https://gitlab.torproject.org/tpo/tpa/status-site/-/merge_requests/71/diffs and https://review.torproject.net/tpo/tpa/status-site/review-v6-y8ucnr/ 

i like how the new v6 looks though, congrats! :)

**Comment by mistermantas** [comment]
> i think i don't understand this fix: looking at [ad0c670](https://github.com/cstate/cstate/commit/ad0c67039be177e60dade470a4c5d8ffe117ba98), it refers to this issue, but doesn't seem related to the calendar stuff...
> 
> i've also tried adding a "calendar" output to the "home" output and it didn't change much... see https://gitlab.torproject.org/tpo/tpa/status-site/-/merge_requests/71/diffs and https://review.torproject.net/tpo/tpa/status-site/review-v6-y8ucnr/
> 
> i like how the new v6 looks though, congrats! :)

The calendar stuff wasn't implemented with an export but a JS button on the page that generates an .ics

```
	{{/* Only show the button if it's informational or a "notice" event */}}
	{{ if or .Params.informational (eq .Params.severity "notice") }}

	<div class="padding"></div>
	<p><small class="faded"><b>Add this period to your calendar</b></small></p>
	<button id="addToCalendarBtn" class="tag" style="padding: 8px 12px">
		Download event (.ics)
	</button>
```

**Comment by anarcat** [comment]
okay, so it only shows up for notices, is that it?

because i couldn't see it directly in our setup...

i was expecting a single ics endpoint one would subscribe to for *any*
scheduled (or past!) outages...

**Comment by mistermantas** [comment]
Currently it's limited that way but I can update the upstream to be more fully featured - even this implementation happens to be a bit buggy and the generated .ics file looks malformed in certain cases

It just seems weird to have an "add to calendar" button for downtime/service disruption if you don't know when it'll end?

Theoretically if it was possible to have a subscription feed (which I think is what you're describing, and NOT .ics invites), then all incidents (mostly past incidents and future incidents with set end dates) could be queried.

**Comment by anarcat** [comment]
> Theoretically if it was possible to have a subscription feed (which I think is what you're describing, and NOT .ics invites), then all incidents (mostly past incidents and future incidents with set end dates) could be queried. 

Yeah, essentially my thinking is there should be a single ICS file or
*something* (but i think ICS files can work for this no?) that has *all*
the events in there, past and future.

Maybe that's WebCal or CalDAV and that requires server side support?

**Comment by jcharaoui** [comment]
In Thunderbird, at least, you can definitely "subscribe" to an .ICS calendar using a web-based URL, which it will automatically and  periodically query and keep up to date, see https://support.mozilla.org/en-US/kb/adding-a-holiday-calendar#w_subscribe-to-it-on-the-internet
